rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,337 @@
+
+```go
+// == The `Config` struct is sent with the `init` message at the start of the protocol.
+
+// The following old fields have been completely removed:
+//   - `metricsPort` (moving over to push-based stats syncing, where we will sync any metrics we want to expose)
+//   - `unsafeNotTrustIp` (only used as a hack in old integration test framework; having it makes p2p code harder to reason about)
+//   - `gaterConfig` (moving towards more abstracted interface in which Go manages gating state data)
+type Config struct {
+  networkId           string      // unique network identifier
+  privateKey          string      // libp2p id private key
+  stateDirectory      string      // directory to store state in (peerstore and dht will be stored/loaded from here)
+  listenOn            []Multiaddr // interfaces we listen on
+  externalAddr        Multiaddr   // interface we advertise for other nodes to connect to
+  floodGossip         bool        // enables gossip flooding (should only be turned on for protected nodes hidden behind a sentry node)
+  directPeers         []Multiaddr // forces the node to maintain connections with peers in this list (typically only used for sentry node setups and other specific networking scenarios; these peers are automatically trustlisted)
+  seedPeers           []Multiaddr // list of seed peers to connect to initially (seeds are automatically trustlisted)
+  maxConnections      int         // maximum number of connections allowed before the connection manager begins trimming open connections
+  validationQueueSize int         // size of the queue of active pending validation messages
+  // TODO: peerExchange bool vs minaPeerExchange bool (seems like at least one of these should be deprecated)
+  //   - peerExchange == enable libp2p's concept of peer exchange in the pubsub options
+  //   - minaPeerExchange == write random peers to connecting peers
+}
+
+// == The `Stats` struct is sent on an interval via the `stats` message.
+// == It contains metrics and statistics relevant to the helper process,
+// == to be further exposed by the daemon as prometheus metrics.
+
+type Stats struct {
+  storedPeerCount       int
+  connectedPeerCount    int
+  messageMinSize        float64
+  messageMaxSize        float64
+  messageAvgSize        float64
+  latencyMin            float64
+  latencyMax            float64
+  latencyAvg            float64
+  totalBandwidthIn      float64
+  totalBandwidthOut     float64
+  totalBandwidthRateIn  float64
+  totalBandwidthRateOut float64
+}
+
+// == A `ValidationStatus` notifies the helper process of whether or not
+// == a message was valid (or relevant).
+
+type ValidationStatus int
+const (
+  VALIDATION_ACCEPT ValidationStatus = iota
+  VALIDATION_REJECT
+  VALIDATION_IGNORE
+)
+
+// == These types define the concept of an `AbstractPeerGraph`, which
+// == describes a peer traversal algorithm for the helper to perform
+// == as when finding a successful response to an RPC query.
```

Comment:
Can you describe more how this graph is interpreted when sending a request? This is to prevent round-trips to ocaml on retries right? What is the representation for the "start" node of the graph (where we explore for the initial query) -- what does it mean if a node is connected to another in the graph but not this start node? Is it a representation of "if this request fails/times out, then request the connecting node?